src/tango/ip/fd_ip.c

```diff
@@ -0,0 +1,381 @@
+  arp->hw_type        = htons( 1 );            /* Ethernet is 1 */
+  arp->proto_type     = htons( 0x0800 );       /* IP is 0x0800 */
+  arp->hw_addr_len    = 6;                     /* hardware address length - ethernet is 6 */
+  arp->proto_addr_len = 4;                     /* protocol address length - IPv4 is 4 */
+  arp->op             = htons( 1 );            /* operation - request is 1 */
+
+  fd_memcpy( arp->sender_hw_addr,    src_mac_addr,     6 ); /* sender hardware address */
+  fd_memcpy( arp->sender_proto_addr, &net_src_ip_addr, 4 ); /* sender protocol (IPv4) address */
+
+  fd_memset( arp->target_hw_addr,    0,                6 ); /* target hardware address - ignored for request */
+  fd_memcpy( arp->target_proto_addr, &net_dst_ip_addr, 4 ); /* target protocol (IPv4) address - ignored for request */
+
+  if( arp_len ) *arp_len = sizeof( *arp );
+
+  return 0;
+}
+
+
+/* fetch the routing table from the kernel
+
+   the routing table will be written into the workspace, completely replacing
+   any existing routing entries */
+
+void
+fd_ip_route_fetch( fd_ip_t * ip ) {
+  fd_ip_route_entry_t * route_table     = fd_ip_route_table_get( ip );
+  ulong                 route_table_cap = ip->num_route_entries;
+  fd_nl_t *             netlink         = fd_ip_netlink_get( ip );
+
+  long num_entries = fd_nl_load_route_table( netlink, route_table, route_table_cap );
+
+  if( num_entries < 0L ) {
+    /* TODO don't switch */
+    return;
+  }
+
+  /* TODO switch */
+  ip->cur_num_route_entries = (ulong)num_entries;
+}
+
+/* query the routing table
+
+   the provided IP address is looked up in the routing table
+
+   if an appropriate entry is found, the details are written into
+     the destination and 0 is returned
+
+   otherwise, 1 is returned */
+
+int
+fd_ip_route_query( fd_ip_t *              ip,
+                   fd_ip_route_entry_t ** route,
+                   uint                   ip_addr ) {
+  fd_ip_route_entry_t * route_table     = fd_ip_route_table_get( ip );
+  ulong                 route_table_cap = ip->num_route_entries;
+
+  fd_ip_route_entry_t * entry = fd_nl_route_query( route_table, route_table_cap, ip_addr );
+
+  if( FD_UNLIKELY( !entry ) ) return 1;
+
+  *route = entry;
+
+  return 0;
+}
+
+
+int
+fd_ip_route_ip_addr( uchar *   out_dst_mac,
+                     uint *    out_next_ip_addr,
+                     uint *    out_ifindex,
+                     fd_ip_t * ip,
+                     uint      ip_addr ) {
+  /* handle broadcasts and multicast */
+
+  /* multicast is 224.0.0.0/4 */
+  if( ( ip_addr & 0xf0000000 ) == 0xe0000000 ) {
+    /* multicast */
+
+    /* map to ethernet space */
+    out_dst_mac[0] = 0x01;
+    out_dst_mac[1] = 0x00;
+    out_dst_mac[2] = 0x5e;
+    out_dst_mac[3] = (uchar)( ( ip_addr >> 020 ) & 0x7fU );
+    out_dst_mac[4] = (uchar)( ( ip_addr >> 010 ) & 0xffU );
+    out_dst_mac[5] = (uchar)( ( ip_addr >> 000 ) & 0xffU );
+
+    /* TODO choose an ifindex */
```

Comment:
Noting the TODO. What's a good strategy? Send to same ifindex that routes to 1.1.1.1 or index 0 if not found?

Reply:
I guess with sockets, broadcast and multicast are either bound to an interface, or will use ALL multicast or broadcast interfaces available respectively. 
So the routing algo really doesn't need to do anything here.